Simplenote/SPHistoryLoader.swift

```diff
@@ -0,0 +1,109 @@
+import Foundation
+
+// MARK: - SPHistoryLoader: Request and aggregate versions of Simperium object
+//
+final class SPHistoryLoader {
+
+    // MARK: - Item: Represents a version object
+    //
+    struct Item {
+        let version: Int
+        let data: [String: Any]
+    }
+
+    private let bucket: SPBucket
+    private let simperiumKey: String
+    private let amountOfVersionsToLoad: Int
+
+    private var completion: (([Item]) -> Void)?
+    private var data: [Int: [String: Any]] = [:]
+    private var sortedItems: [Item] {
+        data.map({ Item(version: $0, data: $1) })
+            .sorted(by: { $0.version < $1.version })
+    }
+
+    /// Designated Initializer
+    ///
+    /// - Parameters:
+    ///     - bucket: Simperium bucket
+    ///     - simperiumKey: Key of Object for which to load history
+    ///     - currentVersion: Current version of Object. Used to calculate the amount of available versions to load
+    ///
+    init(bucket: SPBucket, simperiumKey: String, currentVersion: Int) {
+        self.bucket = bucket
+        self.simperiumKey = simperiumKey
+        amountOfVersionsToLoad = min(currentVersion, Constants.maxNumberOfVersions)
+    }
+
+    /// Load verions
+    ///
+    /// - Parameters:
+    ///     - completion: Invoked after all requested versions have arrived
+    ///
+    func load(completion: @escaping ([Item]) -> Void) {
+        if self.completion != nil {
+            return
+        }
+
+        data = [:]
+        self.completion = completion
+
+        bucket.requestVersions(Int32(amountOfVersionsToLoad), key: simperiumKey)
+    }
+}
+
+// MARK: - Data processing
+//
+extension SPHistoryLoader {
+    /// Process and store a version
+    ///
+    /// As Simperium supports only one delegate and AppDelegate is set as a delegate, so some other
+    /// class will pass data to SPHistoryLoader. (In case of note version, it's note editor)
+    ///
+    /// - Parameters:
+    ///     - data: data of this version
+    ///     - version: version of an object
+    ///
+    func process(data: [String: Any], forVersion version: Int) {
+        self.data[version] = data
+        checkIfFinished()
+    }
+
+    private func checkIfFinished() {
+        guard data.count == amountOfVersionsToLoad else {
+            return
```

Comment:
Question: If, for whatever reasons, not all of the requested versions make it back (networking, new document?), would this API get stuck?

Reply:
I think it may stuck, but...
Now we do:
* show history only if current version of the note > 1, which should mean that it's not a new document anymore
* in case simperium authenticator is not connected when we're about to fetch versions, error message is shown and history loader is not invoked
* I see that web socket reconnects automatically on errors and should resume fetching versions. So in case of some brief network error things should get resolved
* User can always dismiss the history view (in case loading takes to long or stuck for some reason)

Maybe we can:
* Set a timeout and show a message if loading takes too long
* Listen to reachability notifications and update UI in some way to reflect possible connectivity issues

What do you think?

Reply:
(For the record!) replying in a comment, to "unthread" things up a bit!